contracts/adapters/CharmAdapter.sol

```diff
@@ -0,0 +1,388 @@
+      //transfer over
+      market.baseToken().uniTransfer(payable(recipient), amountOut);
+
+      emit Exercised(
+          msg.sender,
+          options,
+          optionID,
+          amount,
+          amountOut
+      );
+    }
+
+    function createShort(ProtocolAdapterTypes.OptionTerms memory, uint256)
+        external
+        pure
+        override
+        returns (uint256)
+    {
+        return 0;
+    }
+
+    function closeShort() external pure override returns (uint256) {
+        return 0;
+    }
+
+    /**
+     * @notice Function to lookup oToken addresses.
+     * @param optionTerms is the terms of the option contract
+     */
+    function lookupOToken(ProtocolAdapterTypes.OptionTerms memory optionTerms)
+        public
+        view
+        returns (address token)
+    {
+        bool isPut =
+            optionTerms.optionType == ProtocolAdapterTypes.OptionType.Put;
+
+        //market.baseToken() is underlying asset if call. Strike currency if put. Represents ETH if equal to 0x0
+        address underlying = isPut == true ? optionTerms.strikeAsset : optionTerms.underlying;
+
+        //there should only be a collateral asset for charm if we are writing an option
+        bool isShort = optionTerms.collateralAsset == address(1) ? false : true;
+
+        // refer to mapping after encoding with _getOptionId
+        bytes32 id = _getOptionId(underlying, isShort, optionTerms.strikePrice, optionTerms.expiry, isPut);
+
+        token = idToAddress[id];
+    }
+
+    // Populate mappings of option type to option token address
+    function populateOTokenMappings(ProtocolAdapterTypes.OptionTerms memory optionTerms)
+        internal
+    {
+
+      bool isPut =
+          optionTerms.optionType == ProtocolAdapterTypes.OptionType.Put;
+
+      //market.baseToken() is underlying asset if call. Strike currency if put. Represents ETH if equal to 0x0
+      address underlying = isPut == true ? optionTerms.strikeAsset : optionTerms.underlying;
+
+      uint256 i = 0;
+
+      address[] memory markets = optionFactory.markets();
+
+      while (i < markets.length) {
+        IOptionMarket market = IOptionMarket(markets[i]);
+
+        bool isMarketPut = market.isPut();
+        uint256 marketExpiry = market.expiryTime();
+        IERC20 baseToken = market.baseToken();
+
+        if(!market.isExpired() && underlying == address(baseToken) && isPut == isMarketPut && optionTerms.expiry == marketExpiry){
+          populateMarket(market, baseToken, isMarketPut, marketExpiry);
+          break;
+        }
+
+        i+=1;
+      }
+    }
+
+    function populateMarket(
+      IOptionMarket market,
+      IERC20 baseToken,
+      bool isPut,
+      uint256 expiry
+    ) internal {
+        IOptionToken[] memory longTokens = market.longTokens();
+        IOptionToken[] memory shortTokens = market.shortTokens();
+        uint256[] memory strikePrices = market.strikePrices();
+
+        uint256 j = 0;
+        while(j < longTokens.length){
```

Comment:
Same idea here https://github.com/ribbon-finance/structured-products/pull/86/files#r592965279

Reply:
what do you mean by this? same logic of going in reverse doesn't really apply here because either all long/short tokens are populated with a market or none...

Reply:
Gotcha, doesn't make sense